### PR TITLE
Change MungedPackageName to not be opaque

### PR DESCRIPTION
**Pull request by phadej**
See commit messages.

`cabal-install` is not done. Asking for the review now already. I think `cabal-install` part should be quite straight-forward fixing of type errors

### CODE REVIEW

File: Cabal/Distribution/Backpack/ReadyComponent.hs [hunk 1]
```diff
@@ -275,7 +272,7 @@ toReadyComponents pid_map subst0 comps
                             fmap rc_munged_id (join (Map.lookup dep_uid s)))]
                   where
                     err_pid = MungedPackageId
-                        (mkMungedPackageName "nonexistent-package-this-is-a-cabal-bug")
+                        undefined -- (mkMungedPackageName "nonexistent-package-this-is-a-cabal-bug")
```
Comment (phadej) [inline Cabal/Distribution/Backpack/ReadyComponent.hs]: oops. Will fix.

**Comment by phadej** [comment]
Travis is green. Review would be highly appreciated.

File: Cabal/Distribution/Types/InstalledPackageInfo/FieldGrammar.hs [hunk 2]
```diff
@@ -253,14 +265,17 @@ basicPkgName f b = (\x -> b { _basicPkgName = x }) <$> f (_basicPkgName b)
...
     :: (FieldGrammar g, Applicative (g Basic))
     => g Basic Basic
-basicFieldGrammar = Basic
+basicFieldGrammar = mkBasic
     <$> optionalFieldDefAla "name"          MQuoted  basicName (mungedPackageName emptyInstalledPackageInfo)
     <*> optionalFieldDefAla "version"       MQuoted  basicVersion nullVersion
     <*> optionalField       "package-name"           basicPkgName
     <*> optionalField       "lib-name"               basicLibName
+  where
+    mkBasic n v pn ln = Basic n v pn (maybe LMainLibName LSubLibName ln)
```
Comment (fgaz) [inline Cabal/Distribution/Types/InstalledPackageInfo/FieldGrammar.hs]: `basicLibName` should be of type `LibraryName`

File: Cabal/Cabal.cabal [hunk 3]
```diff
@@ -370,6 +370,7 @@ library
     Distribution.Types.LibraryName
     Distribution.Types.MungedPackageName
     Distribution.Types.PackageName
+    Distribution.Types.PackageName.Magic
```
[the remark this answers is not in the document]
Comment (phadej) [reply Cabal/Cabal.cabal]: Not easily. Won't do.

File: Cabal/Distribution/Types/InstalledPackageInfo/FieldGrammar.hs (hunk 2 above)
Comment (phadej) [reply Cabal/Distribution/Types/InstalledPackageInfo/FieldGrammar.hs]: it's not exported. wont'fix. (Also `LibraryName` doesn't have `Parsec` instance on purpose, so I'd need to introduce `newtype` etc.)

File: Cabal/Distribution/Simple/Configure.hs [hunk 4]
```diff
@@ -1230,7 +1229,7 @@ selectDependency pkgid internalIndex installedIndex requiredDepsMap
   -- even if there is a newer installed library "MyLibrary-0.2".
   case Map.lookup dep_pkgname internalIndex of
     Just cname -> if use_external_internal_deps
-                    then do_external (Just cname) <$> Set.toList libs
+                    then do_external (Just $ maybeToLibraryName cname) <$> Set.toList libs
```
[the remark this answers is not in the document]
Comment (phadej) [reply Cabal/Distribution/Simple/Configure.hs]: Won't fix. Properly done current `Map PackageName (Maybe UnqualComponentName)` into a **newtype** over `Map PackageName LibraryName`

File: Cabal/Distribution/Types/InstalledPackageInfo/FieldGrammar.hs (hunk 2 above)
Comment (phadej) [reply Cabal/Distribution/Types/InstalledPackageInfo/FieldGrammar.hs]: Maybe even using `InstalledPackageIndex` would make sense, even it's not "installed", but "package index" anyway.

**Pull request merged**